src/dbnode/storage/index.go

```diff
@@ -2066,6 +2066,114 @@ func (i *nsIndex) CleanupExpiredFileSets(t time.Time) error {
 	return i.deleteFilesFn(filesets)
 }
 
+func (i *nsIndex) CleanupCorruptedFileSets() error {
+	/*
+	   Corrupted index filesets can be safely cleaned up if its not
+	   the latest volume index per index volume type/block start combo.
+
+	   We are guaranteed not to be actively writing to an index fileset once
+	   we're already writing to later volume indices.
+	*/
+	fsOpts := i.opts.CommitLogOptions().FilesystemOptions()
+	infoFiles := i.readIndexInfoFilesFn(fs.ReadIndexInfoFilesOptions{
+		FilePathPrefix:   fsOpts.FilePathPrefix(),
+		Namespace:        i.nsMetadata.ID(),
+		ReaderBufferSize: fsOpts.InfoReaderBufferSize(),
+		IncludeCorrupted: true,
+	})
+
+	if len(infoFiles) == 0 {
+		return nil
+	}
+
+	var (
+		toDelete []string
+		begin    = 0 // marks the beginning of a subslice that contains filesets with same block starts
+	)
+	// It's expected that info files are ordered by block start and volume index
+	for j := range infoFiles {
+		if infoFiles[begin].ID.BlockStart.Before(infoFiles[j].ID.BlockStart) {
+			files, err := i.getCorruptedVolumesForDeletion(infoFiles[begin:j])
+			if err != nil {
+				return err
+			}
+			toDelete = append(toDelete, files...)
+			begin = j
+		} else if infoFiles[begin].ID.BlockStart.After(infoFiles[j].ID.BlockStart) {
+			errorMessage := "filesets are expected to be ordered by block start"
+			instrument.EmitAndLogInvariantViolation(i.opts.InstrumentOptions(), func(l *zap.Logger) {
+				l.Error(errorMessage)
+			})
+			return instrument.InvariantErrorf(errorMessage)
+		}
+	}
+
+	// Process the volumes in the last block, which are not covered by the loop.
+	files, err := i.getCorruptedVolumesForDeletion(infoFiles[begin:])
+	if err != nil {
+		return err
+	}
+	toDelete = append(toDelete, files...)
+
+	return i.deleteFilesFn(toDelete)
+}
+
+func (i *nsIndex) getCorruptedVolumesForDeletion(filesets []fs.ReadIndexInfoFileResult) ([]string, error) {
+	if len(filesets) <= 1 {
+		return nil, nil
+	}
+
+	// Check for invariants.
+	for j := 0; j+1 < len(filesets); j++ {
```

Comment:
Nit: I think this loop would be more readable in this form:
```suggestion
	for j := 1; j < len(filesets); j++ {
```
And then adjusting the indexes in the loop body accordingly.